include/seqan3/alphabet/cigar/cigar.hpp

```diff
@@ -0,0 +1,326 @@
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Contains cigar composition alphabet.
+ * \author Svenja Mehringer <svenja.mehringer AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp>
+#include <seqan3/alphabet/cigar/cigar_op.hpp>
+#include <seqan3/alphabet/composition/cartesian_composition.hpp>
+#include <seqan3/std/concepts>
+#include <seqan3/std/ranges>
+
+#include <range/v3/view/zip.hpp>
+
+namespace seqan3
+{
+
+/*!\brief The cigar composition joins the seqan3::cigar_op alphabet with a count value.
+ * \ingroup cigar
+ * \tparam cigar_type Must be of type seqan3::cigar_op.
+ * \tparam count_type Type of the count value; must model std::UnsignedIntegral.
+ *
+ * This alphabet can be used to represent an alignment in a compressed way using
+ * the CIGAR string representation. The seqan3::cigar alphabet implements the
+ * CIGAR elements in this context consisting of a CIGAR operator (seqan3::cigar_op)
+ * and a count value.
+ *
+ * Example usage:
+ * \snippet test/snippet/alphabet/cigar/cigar_op.cpp general
+ *
+ * \note Usually you do not want to manipulate cigar elements and vectors on
+ *       your own but convert an alignment to a cigar and back. See
+ *       seqan3::get_cigar_vector for how to convert two aligned sequences into
+ *       an cigar_vector.
+ */
+template<typename cigar_type = cigar_op, typename count_type = uint32_t>
```

Comment:
I would switch the order. First, cigar_type is kind of fixed, albeit we need to declare it as a template in order to inherit the tuple size, tuple element functions. I am wondering if we should drop it and reimplement the functions specifically for the cigar alphabet. Because it is kind of misleading to make the interface generic while it is not generic. It does not justify to save 6 lines of code. Unless it would be more complicated? Second, the user might want to use a smaller int type and `cigar<uint16_t>` is better than `cigar<cigar_op, uint16_t>`

Reply:
I see your point for switching the template arguments but I think there is more than 6 lines of code to change the cigar alphabet to its own composition. The cartesian_composition base class includes 600 lines of codes that would need to be copy and pasted, since the functionality for assigning, constructing and comparing a compound type is redundant (e.g. `cigar<> c = 4;` sets only the count value). So the question is if it justifies 300 lines of redundant code? I'm not sure myself..

Reply:
No, I probably didn't express myself correctly.
The cigar should still be a composition and inherit from it. But my last understanding, was, that the std::tuple_element and std::tuple_size wouldn't work if cigar does not offer the exact same template arguments as the std::tuple it is inheriting from through the composition. But you could overload those two meta-functions for your cigar class. Still the other inherited functions should work.

Reply:
Ah I see now! I will check if this is the only 6 lines of code we would need to overload. If so, I agree to change it.

Reply:
Done.